### PR TITLE
GG-39400 Do not complete foreign connection futures when failing to reserve recovery descriptor

### PR DESCRIPTION
**Pull request by rpuch**
https://ggsystems.atlassian.net/browse/GG-39400

When processing an incoming connection attempt, seeing a future created by another initiator of the same connection and failing to reserve the corresponding connection, we should not complete this future and remove it from the map of such futures. If we do it, we break the 'at most 1 physical connection per connection key' invariant.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java [hunk 1]
```diff
@@ -711,25 +711,19 @@ else if (discoverySpi instanceof IgniteDiscoverySpi)
                     ses.send(new RecoveryLastReceivedMessage(ALREADY_CONNECTED));
                 }
                 else {
-                    try {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java:714]: Why did you remove the try-finally block?

File: modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/GridTcpCommunicationConnectionExclusivenessTest.java [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+    /**
+     * Tests a scenario of ensureNoDuplicatesWhenInboundCompetesWithOutbounds(boolean) where
+     * the order of B is higher than the order of A.
+     *
+     * @throws Exception If failed.
+     * @see #ensureNoDuplicatesWhenInboundCompetesWithOutbounds(boolean)
+     */
+    @Test
+    public void noDuplicatesWhenInboundFromHigherNodeCompetesWithOutbounds() throws Exception {
+        ensureNoDuplicatesWhenInboundCompetesWithOutbounds(true);
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/GridTcpCommunicationConnectionExclusivenessTest.java:117]: Can we use a parameterized test?

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java (hunk 1 above)
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java:714]: Because it was incorrect to always remove `oldFut` from the map. We must only remove it if we managed to create a connection.

Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java:714]: ok, so it was originally there just to save us from code duplication?

File: modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/GridTcpCommunicationConnectionExclusivenessTest.java (hunk 2 above)
Comment (rpuch) [reply modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/GridTcpCommunicationConnectionExclusivenessTest.java]: Rewrote to Parameterized runner

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java (hunk 1 above)
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java:714]: I think it was there *due* to code duplication, as a result of a copy-paste

**Review by sashapolo — APPROVED** [review]
